### PR TITLE
Adding solutions in go

### PR DESCRIPTION
**Pull request by khvr1993**
I will be adding solutions one by one as I go through the questions. This is a work in progress PR

### CODE REVIEW

File: go/01_arrays_and_hashing/665_Non_decreasing_Array/238_Product_of_Array_Except_Self/238_Product_of_Array_Except_Self.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+func productExceptSelf(nums []int) []int {
+	result := make([]int, len(nums))
+	prefixProductArr := getPrefixProduct(nums)
+	postFixProductArr := getPostFixProduct(nums)
+	prefixValue := 1
+	postFixValue := 1
+
+	for i := range result {
```
Comment (Maximilian5189) [inline go/01_arrays_and_hashing/665_Non_decreasing_Array/238_Product_of_Array_Except_Self/238_Product_of_Array_Except_Self.go:32]: I think these if statements could be avoided, for example:
```go
func productExceptSelf(nums []int) []int {
    res := make([]int, len(nums))
    
    prefix := 1
    res[0] = 1
    for i := 0; i < len(nums); i++ {
        res[i] = prefix
        
        prefix *= nums[i]
    }
    
    postfix := 1
    for i := len(nums)-1; i >= 0; i-- {
        res[i] *= postfix
        
        postfix *= nums[i]
    }
    
    return res
}
```

Comment (khvr1993) [reply go/01_arrays_and_hashing/665_Non_decreasing_Array/238_Product_of_Array_Except_Self/238_Product_of_Array_Except_Self.go:32]: Yes agreed. The solution I have written is a suboptimal one where I consume extra space for prefix and postfix array.

The proposed solution uses O(1) space and a cleaner way of doing things. I will add it as a new procedure and write tests. Thanks for the suggestion

**Comment by khvr1993** [comment]
Hi @Maximilian5189  Could you review this and if you have write access request you to merge the PR.
Thank you

**Review by mitchellirvin — CHANGES_REQUESTED** [review]
hi @khvr1993 ! thanks so much for working on all of these problems. 

a couple things:
1. do you mind removing the test files (we currently don't track test files in any other language/area, and it's highly likely they would diverge from the source code over time, though I love your enthusiasm for tests!)
2. do you mind moving all of the solutions to the top level `go/` directory? admittedly this guidance isn't clear, but I think we're going to move to keeping all solutions in the top level dir of each language to make searchability easier
3. do you mind updating the file names to match the existing conventions? details can be found here: https://github.com/neetcode-gh/leetcode#guidelines
4. do you mind breaking this PR up into a few smaller PRs for review? I usually wouldn't recommend doing more than a handful of solutions per-PR, as it increases the time it takes for review and merging, and it decreases the confidence we have that everything we're merging is correct

Once these things are done, I'll go through and review each solution and verify it passes submission on leetcode.com (and that there isn't already an existing solution in the repo for go). we're super grateful for the contributions, but in the future if you keep your PRs smaller, it'll take less time to get it reviewed and merged!

**Comment by khvr1993** [comment]
@mitchellirvin Thanks for the suggestions. 
1. I added test files so that the users could run most of the basic code and try it out themselves. since for Go main package is mandatory to run the code, having tests would make life easy to run sample cases. I will remove them as other languages are not having them
2. Sure
3. Sure
4. I will make the changes and try to make small PRs. I thought I could add solutions in bulk and merge them at once. It makes sense that the reviewer would get overwhelmed.
All the codes are tested in leetcode and are accepted.

**Comment by mitchellirvin** [comment]
@khvr1993 thanks so much! Please link the PRs when they're open and I'll review them as soon as I can. 

I trust that you've submitted them on leetcode and they pass, I just always have to go run them myself just in case!

**Comment by Ahmad-A0** [comment]
I'll close this as stale since the directory structure for go solutions has changed, feel free to update the structure in which case I'll reopen and merge this.

**Pull request closed without merging**